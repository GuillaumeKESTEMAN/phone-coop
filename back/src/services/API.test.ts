import initAPI from './API';
import FULL_CONFIG from '../config/test/config';
import { getOpenAPIOperations } from '@whook/http-router';
import OpenAPISchemaValidator from 'openapi-schema-validator';
import { initAPIDefinitions, initImporter } from '@whook/whook';
import path from 'path';

describe('API', () => {
  const { CONFIG } = FULL_CONFIG;
  const BASE_URL = 'http://localhost:1337';
  const log = jest.fn();
  let API_DEFINITIONS;

  beforeAll(async () => {
    const importer = await initImporter({ log });

    API_DEFINITIONS = await initAPIDefinitions({
      PROJECT_SRC: path.join(__dirname, '..'),
      WHOOK_PLUGINS_PATHS: [path.dirname(require.resolve('@whook/whook/dist'))],
      importer,
    });
  });

  beforeEach(() => {
    log.mockReset();
  });

  it('should work', async () => {
    const API = await initAPI({
      ENV: {},
      CONFIG,
      BASE_URL,
      API_VERSION: '1.1.0',
      API_DEFINITIONS,
      log,
    });

    expect({
      API,
      logCalls: log.mock.calls.filter(([type]) => !type.endsWith('stack')),
    }).toMatchSnapshot();
  });

  it('should always use declared security definition', async () => {
    const API = await initAPI({
      ENV: {},
      CONFIG,
      BASE_URL,
      API_VERSION: '1.1.0',
      API_DEFINITIONS,
      log,
    });
    const securitySchemes = API.components.securitySchemes;
    const operations = getOpenAPIOperations(API);

    expect(
      operations
        .filter((operation) => operation.security && operation.security.length)
        .filter((operation) =>
          operation.security.some((operationSecurity) =>
            Object.keys(operationSecurity).some(
              (operationSecurityName) =>
                !securitySchemes[operationSecurityName],
            ),
          ),
        )
        .map(({ method, path }) => `${method} ${path}`)
        .sort(),
    ).toEqual([]);
  });

  it('should produce a valid OpenAPI file', async () => {
    const API = await initAPI({
      ENV: {},
      CONFIG,
      BASE_URL,
      API_VERSION: '1.1.0',
      API_DEFINITIONS,
      log,
    });

    const result = new OpenAPISchemaValidator({ version: 3 }).validate(
      // Temporar type fix due to version mismatch of OpenAPIV3
      // between Whook and OpenAPISchemaValidator
      API as any,
    );

    expect({ result }).toMatchSnapshot();
  });

  describe('should always have the same amount of', () => {
    let operations;

    beforeAll(async () => {
      const API = await initAPI({
        ENV: {},
        CONFIG,
        BASE_URL,
        API_VERSION: '1.1.0',
        API_DEFINITIONS,
        log,
      });

      operations = getOpenAPIOperations(API);
    });

    it('endpoints', async () => {
      expect(
        operations
          .filter(
            (operation) =>
              !operation['x-whook'] || !operation['x-whook'].private,
          )
          .map(({ method, path }) => `${method} ${path}`)
          .sort(),
      ).toMatchSnapshot();
    });

    it('publicly documented endpoints', async () => {
      expect(
        operations
          .filter(
            (operation) =>
              !operation['x-whook'] || !operation['x-whook'].private,
          )
          .map(({ method, path }) => `${method} ${path}`)
          .sort(),
      ).toMatchSnapshot();
    });

    it('non authenticated endpoints', async () => {
      expect(
        operations
          .filter(
            (operation) =>
              !operation.security || operation.security.length === 0,
          )
          .map(({ method, path }) => `${method} ${path}`)
          .sort(),
      ).toMatchSnapshot();
    });

    it('optionally authenticated endpoints', async () => {
      expect(
        operations
          .filter(
            (operation) =>
              operation.security &&
              operation.security.some(
                (security) => Object.keys(security).length === 0,
              ),
          )
          .map(({ method, path }) => `${method} ${path}`)
          .sort(),
      ).toMatchSnapshot();
    });

    it('basic authenticated endpoints', async () => {
      expect(
        operations
          .filter(
            (operation) =>
              operation.security &&
              operation.security.some((security) => security.basicAuth),
          )
          .map(({ method, path }) => `${method} ${path}`)
          .sort(),
      ).toMatchSnapshot();
    });

    it('bearer authenticated endpoints', async () => {
      expect(
        operations
          .filter(
            (operation) =>
              operation.security &&
              operation.security.some((security) => security.bearerAuth),
          )
          .map(({ method, path }) => `${method} ${path}`)
          .sort(),
      ).toMatchSnapshot();
    });
  });
});
