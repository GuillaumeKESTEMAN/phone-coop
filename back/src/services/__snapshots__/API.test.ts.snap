// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API should always have the same amount of basic authenticated endpoints 1`] = `Array []`;

exports[`API should always have the same amount of bearer authenticated endpoints 1`] = `
Array [
  "get /openAPI",
]
`;

exports[`API should always have the same amount of endpoints 1`] = `
Array [
  "get /openAPI",
  "get /phones/brands",
  "get /ping",
]
`;

exports[`API should always have the same amount of non authenticated endpoints 1`] = `
Array [
  "get /phones/brands",
  "get /ping",
  "options /openAPI",
  "options /phones/brands",
  "options /ping",
]
`;

exports[`API should always have the same amount of optionally authenticated endpoints 1`] = `
Array [
  "get /openAPI",
]
`;

exports[`API should always have the same amount of publicly documented endpoints 1`] = `
Array [
  "get /openAPI",
  "get /phones/brands",
  "get /ping",
]
`;

exports[`API should produce a valid OpenAPI file 1`] = `
Object {
  "result": Object {
    "errors": Array [],
  },
}
`;

exports[`API should work 1`] = `
Object {
  "API": Object {
    "components": Object {
      "headers": Object {},
      "parameters": Object {},
      "requestBodies": Object {},
      "responses": Object {},
      "schemas": Object {},
      "securitySchemes": Object {
        "bearerAuth": Object {
          "description": "Bearer authentication with a user API token",
          "scheme": "bearer",
          "type": "http",
        },
      },
    },
    "info": Object {
      "description": "A project to recycle phones",
      "title": "phone-coop-back",
      "version": "1.1.0",
    },
    "openapi": "3.0.3",
    "paths": Object {
      "/openAPI": Object {
        "get": Object {
          "operationId": "getOpenAPI",
          "parameters": Array [],
          "responses": Object {
            "200": Object {
              "content": Object {
                "application/json": Object {
                  "schema": Object {
                    "type": "object",
                  },
                },
              },
              "description": "Provides the private Open API documentation",
            },
          },
          "security": Array [
            Object {},
            Object {
              "bearerAuth": Array [
                "admin",
              ],
            },
          ],
          "summary": "Get API documentation.",
          "tags": Array [
            "system",
          ],
          "x-whook": Object {
            "private": false,
          },
        },
        "options": Object {
          "operationId": "optionsWithCORS",
          "parameters": Array [
            Object {
              "in": "query",
              "name": "access_token",
              "schema": Object {
                "type": "string",
              },
            },
          ],
          "responses": Object {
            "200": Object {
              "description": "CORS sent.",
            },
          },
          "summary": "Enable OPTIONS for CORS",
          "tags": Array [
            "CORS",
          ],
          "x-whook": Object {
            "private": true,
            "sourceOperationId": "getOpenAPI",
            "suffix": "CORS",
            "type": "http",
          },
        },
      },
      "/phones/brands": Object {
        "get": Object {
          "operationId": "getPhoneBrands",
          "parameters": Array [],
          "responses": Object {
            "200": Object {
              "content": Object {
                "application/json": Object {
                  "schema": Object {
                    "items": Object {
                      "properties": Object {
                        "id": Object {
                          "minimum": 1,
                          "type": "number",
                        },
                        "label": Object {
                          "type": "string",
                        },
                      },
                      "required": Array [
                        "id",
                        "label",
                      ],
                      "type": "object",
                    },
                    "type": "array",
                  },
                },
              },
              "description": "Success",
            },
          },
          "summary": "Provide a list of phones brands",
          "tags": Array [],
        },
        "options": Object {
          "operationId": "optionsWithCORS",
          "parameters": Array [],
          "responses": Object {
            "200": Object {
              "description": "CORS sent.",
            },
          },
          "summary": "Enable OPTIONS for CORS",
          "tags": Array [
            "CORS",
          ],
          "x-whook": Object {
            "private": true,
            "sourceOperationId": "getPhoneBrands",
            "suffix": "CORS",
            "type": "http",
          },
        },
      },
      "/ping": Object {
        "get": Object {
          "operationId": "getPing",
          "responses": Object {
            "200": Object {
              "content": Object {
                "application/json": Object {
                  "schema": Object {
                    "additionalProperties": false,
                    "properties": Object {
                      "pong": Object {
                        "enum": Array [
                          "pong",
                        ],
                        "type": "string",
                      },
                    },
                    "type": "object",
                  },
                },
              },
              "description": "Pong",
            },
          },
          "summary": "Checks API's availability.",
          "tags": Array [
            "system",
          ],
        },
        "options": Object {
          "operationId": "optionsWithCORS",
          "parameters": Array [],
          "responses": Object {
            "200": Object {
              "description": "CORS sent.",
            },
          },
          "summary": "Enable OPTIONS for CORS",
          "tags": Array [
            "CORS",
          ],
          "x-whook": Object {
            "private": true,
            "sourceOperationId": "getPing",
            "suffix": "CORS",
            "type": "http",
          },
        },
      },
    },
    "servers": Array [
      Object {
        "url": "http://localhost:1337",
      },
    ],
    "tags": Array [
      Object {
        "name": "system",
      },
    ],
  },
  "logCalls": Array [
    Array [
      "debug",
      "ðŸ¦„ - Initializing the API service!",
    ],
  ],
}
`;
